File: autogen_agent_interface/server/utils/cognitive_bridge.ts
/**
 * Cognitive Bridge - Ponte entre TypeScript e Sistema Cognitivo Python
 * Integra o ANIMA Cognitive Core com o sistema TypeScript existente
 */

import { spawn } from "child_process";
import * as path from "path";
import * as fs from "fs";
import { fileURLToPath } from "url";
import { dirname } from "path";

// Suporte para __dirname em módulos ES
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

interface CognitiveResult {
  enriched_message: string;
  original_task: string;
  confidence: number;
  emotional_tone: string;
  modulation_factors: Record<string, number>;
  reflection: {
    reflection_type: string;
    confidence: number;
    confidence_level: string;
    should_revise: boolean;
    revision_suggestions: string[];
    insights: string[];
  } | null;
  relevant_memories: {
    episodic: Array<{
      task: string;
      success: boolean;
      emotion: string;
      emotion_value: number;
      timestamp: number;
    }>;
    semantic: Array<{
      concept: string;
      definition: string;
      usage_count: number;
      confidence: number;
    }>;
    affective: Array<{
      event: string;
      emotion: string;
      emotion_value: number;
      timestamp: number;
    }>;
  };
  decision: {
    action: string;
    approach: string;
    parameters: Record<string, any>;
  };
  cognitive_summary: Record<string, any>;
}

interface CognitiveContext {
  task: string;
  context?: Record<string, any>;
  user_feedback?: string;
  user_id?: string;
}

/**
 * Processa tarefa com sistema cognitivo Python
 * Retorna contexto cognitivo enriquecido
 */
export async function processWithCognitiveSystem(
  task: string,
  context?: Record<string, any>,
  user_id?: string
): Promise<CognitiveResult | null> {
  try {
    // Verificar se o sistema cognitivo está disponível
    const cognitiveScriptPath = path.join(
      __dirname,
      "../../../anima/orchestrator/cognitive_bridge.py"
    );

    if (!fs.existsSync(cognitiveScriptPath)) {
      console.warn("[CognitiveBridge] Sistema cognitivo não disponível, pulando...");
      return null;
    }

    // Criar contexto cognitivo
    const cognitiveContext: CognitiveContext = {
      task,
      context: context || {},
      user_id: user_id || "default_user",
    };

    // Chamar sistema cognitivo Python via script
    return await callCognitivePython(cognitiveContext);
  } catch (error) {
    console.error("[CognitiveBridge] Erro ao processar com sistema cognitivo:", error);
    return null;
  }
}

/**
 * Chama script Python do sistema cognitivo
 */
async function callCognitivePython(
  context: CognitiveContext
): Promise<CognitiveResult> {
  return new Promise((resolve, reject) => {
    const pythonScript = path.join(
      __dirname,
      "../../../anima/orchestrator/cognitive_bridge.py"
    );

    const pythonProcess = spawn("python", [pythonScript, JSON.stringify(context)], {
      cwd: path.join(__dirname, "../../.."),
      env: { ...process.env, PYTHONUNBUFFERED: "1" },
    });

    let stdout = "";
    let stderr = "";

    pythonProcess.stdout.on("data", (data) => {
      stdout += data.toString();
    });

    pythonProcess.stderr.on("data", (data) => {
      stderr += data.toString();
    });

    pythonProcess.on("close", (code) => {
      if (code !== 0) {
        console.error(`[CognitiveBridge] Python script exited with code ${code}`);
        console.error(`[CognitiveBridge] stderr: ${stderr}`);
        reject(new Error(`Python script failed with code ${code}: ${stderr}`));
        return;
      }

      try {
        // Parse JSON response
        const result = JSON.parse(stdout.trim());
        resolve(result);
      } catch (error) {
        console.error("[CognitiveBridge] Erro ao parsear resposta JSON:", error);
        console.error("[CognitiveBridge] stdout:", stdout);
        reject(new Error(`Failed to parse JSON response: ${error}`));
      }
    });

    pythonProcess.on("error", (error) => {
      console.error("[CognitiveBridge] Erro ao executar Python script:", error);
      reject(error);
    });

    // Timeout de 30 segundos
    setTimeout(() => {
      pythonProcess.kill();
      reject(new Error("Cognitive system timeout (30s)"));
    }, 30000);
  });
}

/**
 * Aprende com resposta recebida
 */
export async function learnFromResponse(
  task: string,
  response: string,
  success: boolean = true,
  user_feedback?: string,
  user_id?: string
): Promise<void> {
  try {
    const cognitiveScriptPath = path.join(
      __dirname,
      "../../../anima/orchestrator/cognitive_bridge.py"
    );

    if (!fs.existsSync(cognitiveScriptPath)) {
      return;
    }

    const learnContext = {
      action: "learn",
      task,
      response,
      success,
      user_feedback,
      user_id: user_id || "default_user",
    };

    await callCognitivePython(learnContext as any);
  } catch (error) {
    console.error("[CognitiveBridge] Erro ao aprender com resposta:", error);
    // Não rejeitar, apenas logar erro
  }
}

/**
 * Obtém resumo do estado cognitivo
 */
export async function getCognitiveSummary(user_id?: string): Promise<Record<string, any> | null> {
  try {
    const cognitiveScriptPath = path.join(
      __dirname,
      "../../../anima/orchestrator/cognitive_bridge.py"
    );

    if (!fs.existsSync(cognitiveScriptPath)) {
      return null;
    }

    const summaryContext = {
      action: "summary",
      user_id: user_id || "default_user",
    };

    const result = await callCognitivePython(summaryContext as any);
    return result?.cognitive_summary || null;
  } catch (error) {
    console.error("[CognitiveBridge] Erro ao obter resumo cognitivo:", error);
    return null;
  }
}

/**
 * Enriquece mensagem com contexto cognitivo (modo simplificado sem Python)
 * Usa apenas heurísticas TypeScript quando o sistema Python não está disponível
 */
export function enrichMessageWithCognitiveContext(
  task: string,
  context?: Record<string, any>
): string {
  // Modo simplificado: apenas adiciona contexto básico
  let enrichedMessage = task;

  // Adicionar contexto de memória se disponível
  if (context?.memoryContext === "SIM") {
    enrichedMessage = `[CONTEXTO DE MEMÓRIA DISPONÍVEL]\n${enrichedMessage}`;
  }

  // Adicionar contexto de conversa se disponível
  if (context?.conversationId) {
    enrichedMessage = `[CONVERSA ${context.conversationId}]\n${enrichedMessage}`;
  }

  return enrichedMessage;
}

File: anima/orchestrator/cognitive_bridge.py
"""
Cognitive Bridge - Ponte Python para integração com TypeScript
Script que permite chamar o sistema cognitivo ANIMA a partir de TypeScript
"""
import sys
import json
import logging
from typing import Dict, Any, Optional

# Configurar logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Importar sistema cognitivo
try:
    from anima.orchestrator import CognitiveOrchestrator
    from anima.core import CognitiveCore
    COGNITIVE_SYSTEM_AVAILABLE = True
except ImportError as e:
    logger.warning(f"Sistema cognitivo não disponível: {e}")
    COGNITIVE_SYSTEM_AVAILABLE = False

# Cache de orquestradores por usuário
_orchestrator_cache: Dict[str, CognitiveOrchestrator] = {}


def get_orchestrator(user_id: str = "default_user") -> Optional[CognitiveOrchestrator]:
    """Obtém orquestrador cognitivo para usuário (com cache)"""
    if not COGNITIVE_SYSTEM_AVAILABLE:
        return None
    
    if user_id not in _orchestrator_cache:
        try:
            # Criar orquestrador (sem modelo por enquanto, apenas para processamento cognitivo)
            # O modelo será usado apenas quando necessário
            _orchestrator_cache[user_id] = CognitiveOrchestrator(
                user_id=user_id,
                model_name="gpt-4",  # Placeholder, não será usado se não houver API key
                api_key=None,  # Não usar modelo externo
                base_url=None,
                enable_emotions=True,
                enable_memory=True,
                enable_meta_reasoning=True,
                enable_regulation=True,
            )
        except Exception as e:
            logger.error(f"Erro ao criar orquestrador para usuário {user_id}: {e}")
            return None
    
    return _orchestrator_cache[user_id]


def process_task(context: Dict[str, Any]) -> Dict[str, Any]:
    """Processa tarefa com sistema cognitivo"""
    task = context.get("task", "")
    user_id = context.get("user_id", "default_user")
    context_data = context.get("context", {})
    user_feedback = context.get("user_feedback")
    
    orchestrator = get_orchestrator(user_id)
    if not orchestrator:
        # Retornar resultado vazio se sistema não disponível
        return {
            "enriched_message": task,
            "original_task": task,
            "confidence": 0.5,
            "emotional_tone": "neutral",
            "modulation_factors": {},
            "reflection": None,
            "relevant_memories": {"episodic": [], "semantic": [], "affective": []},
            "decision": {"action": "process", "approach": "standard", "parameters": {}},
            "cognitive_summary": {}
        }
    
    # Processar tarefa
    try:
        result = orchestrator.process_task(
            task=task,
            context=context_data,
            user_feedback=user_feedback
        )
        return result
    except Exception as e:
        logger.error(f"Erro ao processar tarefa: {e}")
        # Retornar resultado básico em caso de erro
        return {
            "enriched_message": task,
            "original_task": task,
            "confidence": 0.5,
            "emotional_tone": "neutral",
            "modulation_factors": {},
            "reflection": None,
            "relevant_memories": {"episodic": [], "semantic": [], "affective": []},
            "decision": {"action": "process", "approach": "standard", "parameters": {}},
            "cognitive_summary": {}
        }


def learn_from_response(context: Dict[str, Any]) -> Dict[str, Any]:
    """Aprende com resposta recebida"""
    task = context.get("task", "")
    response = context.get("response", "")
    success = context.get("success", True)
    user_feedback = context.get("user_feedback")
    user_id = context.get("user_id", "default_user")
    
    orchestrator = get_orchestrator(user_id)
    if not orchestrator:
        return {"status": "error", "message": "Sistema cognitivo não disponível"}
    
    try:
        orchestrator.learn_from_response(
            task=task,
            response=response,
            success=success,
            user_feedback=user_feedback
        )
        return {"status": "success"}
    except Exception as e:
        logger.error(f"Erro ao aprender com resposta: {e}")
        return {"status": "error", "message": str(e)}


def get_summary(context: Dict[str, Any]) -> Dict[str, Any]:
    """Obtém resumo do estado cognitivo"""
    user_id = context.get("user_id", "default_user")
    
    orchestrator = get_orchestrator(user_id)
    if not orchestrator:
        return {"cognitive_summary": {}}
    
    try:
        summary = orchestrator.get_cognitive_summary()
        return {"cognitive_summary": summary}
    except Exception as e:
        logger.error(f"Erro ao obter resumo: {e}")
        return {"cognitive_summary": {}}


def main():
    """Função principal - processa argumentos da linha de comando"""
    if len(sys.argv) < 2:
        print(json.dumps({"error": "No context provided"}))
        sys.exit(1)
    
    try:
        # Parse JSON context
        context_str = sys.argv[1]
        context = json.loads(context_str)
        
        # Determinar ação
        action = context.get("action", "process")
        
        if action == "process":
            result = process_task(context)
        elif action == "learn":
            result = learn_from_response(context)
        elif action == "summary":
            result = get_summary(context)
        else:
            result = {"error": f"Unknown action: {action}"}
        
        # Retornar resultado como JSON
        print(json.dumps(result, indent=2, default=str))
        
    except Exception as e:
        logger.error(f"Erro no cognitive_bridge: {e}")
        print(json.dumps({"error": str(e)}))


if __name__ == "__main__":
    main()

